server/tserver/src/main/java/org/apache/accumulo/tserver/ScanServer.java

```diff
@@ -246,9 +246,30 @@ public ScanServer(ScanServerOpts opts, String[] args) {
         LOG.warn(
             "Tablet metadata caching less than one minute, may cause excessive scans on metadata table.");
       }
-      tabletMetadataCache =
-          Caffeine.newBuilder().expireAfterWrite(cacheExpiration, TimeUnit.MILLISECONDS)
-              .scheduler(Scheduler.systemScheduler()).recordStats().build(tabletMetadataLoader);
+
+      // Get the cache refresh percentage property
+      // Value must be less than 100% as 100 or over would effectively disable it
+      double cacheRefreshPercentage =
+          getConfiguration().getFraction(Property.SSERV_CACHED_TABLET_METADATA_REFRESH_PERCENT);
+      Preconditions.checkArgument(cacheRefreshPercentage < cacheExpiration,
+          "Tablet metadata cache refresh percentage is '%s' but must be less than 1",
+          cacheRefreshPercentage);
+
+      var builder = Caffeine.newBuilder().expireAfterWrite(cacheExpiration, TimeUnit.MILLISECONDS)
```

Comment:
We may be able to supply an executor with:
```
      var builder = Caffeine.newBuilder().expireAfterWrite(cacheExpiration, TimeUnit.MILLISECONDS)
          .scheduler(Scheduler.systemScheduler()).recordStats().executor(context.getScheduledExecutor());
```

Reply:
If a scheduled thread pool is not the correct type, then something like 
```
  Executor refresher = context.threadPools().getPoolBuilder("scan.server.cache.refresh")
              .numCoreThreads(1).numMaxThreads(4).build();
      var builder = Caffeine.newBuilder().expireAfterWrite(cacheExpiration, TimeUnit.MILLISECONDS)
          .scheduler(Scheduler.systemScheduler()).recordStats().executor(refresher);
      
```
would create a thread pool - not sure what the numbers should be.

Reply:
would you want refresh to be included in this doc snippet's limit?

https://github.com/apache/accumulo/blob/e16976c1786649844c4975b9e5404ee43e9745d2/server/tserver/src/main/java/org/apache/accumulo/tserver/ScanServer.java#L229-L231

Reply:
> would create a thread pool - not sure what the numbers should be.

One factor in that, that I do not fully understand ATM is bulk updates w.r.t. to refresh operation in Caffeine. A comment was made on #4544 stating that caffeine refresh may be able to batch updates. If so, then that could work nicely with use an Accumulo batch scanner to refresh lots of tablets metadata in one go and would also generally mean we probably only need few threads for refresh.

Reply:
The expiration should be staggered based on access - so that could point to using a small number of threads?

Reply:
The idea is to aggregate individual refreshes over a time window, perform the bulk operation, and complete the respective `CompletableFuture` with the value for the given key. Caffeine doesn't know it is being batched, which wouldn't help since most refreshes are triggered from distinct lookups (unlike, for instance, getAll). Since it is an async task the cache merely thinks it took longer to execute, the worst case is slightly skewed stats (but custom stats fixes that). It is a little bit of code wrangling if using java.util.concurrent, but very trivial if you can glue together with a reactive stream library.

Reply:
> The idea is to aggregate individual refreshes over a time window, perform the bulk operation, and complete the respective CompletableFuture with the value for the given key.

I see.  So we could do something like have single thread that continually pull refresh work off a queue, then does an Accumulo batch scan to read all data, and then completes futures related to work it pulled off the queue.  The reload pluging to caffeine could push uncompleted futures onto the queue.

Reply:
yes, that would work, though you don't need a dedicated thread to poll. If you lock within the asyncReload method then you can detect if the request is the start of a new batch and schedule a task to run at the end of the time window. Then lock, accumulate the work, unlock, submit it for processing, and complete the pending futures. The lock hold times are a few nanoseconds to coordinate and you can defer to a processing thread pool to allow for parallel batches if helpful (e.g. if a batch size limit).